### DISCUSSION TITLE
Sticky & repeating keys

### DISCUSSION THREAD

**Discussion by dev-zero** [wezterm/wezterm#883] [Q&A] [answered]
Hi there,

I am not entirely sure yet how to report this, but I have the problem when I long-press keys (usually arrows or backspace) they get sticky in the sense that the key gets repeated even though I stopped pressing it. Pressing a different key interrupts it after some delay, as does reaching the beginning or end of line (probably due to the bell event).

A related issue is that after a couple of screen sleeps (activated via screensaver) all keys in a wezterm Window are repeated 3 times. In all the tabs of that window, making that window unusable.

This is all on Wayland with Sway, running `20210502-154244-3f7122cb` (Arch-Linux). Kitty does not exhibit that problem.

Any ideas?
Thanks in advance,
Tiziano

**Comment by wez** [comment, marked as answer]
That is an issue with the smithay-client-toolkit key repeat handling when running under wayland.  (https://github.com/wez/wezterm/issues/669 is tracking that area).

I recommend using XWayland with wezterm for now; that's why `enable_wayland=false` by default.

**Reply by dev-zero** [reply]
Thanks for the quick answer. Switching to X11 backend indeed solved the issues (with the drawback that `rofimoji` is unable to insert characters anymore since it uses `wtype` for that and this doesn't seem to work with XWayland windows).
I couldn't find any issues over at https://github.com/Smithay/client-toolkit/issues, should I report it there?

**Reply by wez** [reply]
No need to report over there: I have a local branch with an alternative key repeat implementation.  That's part of some bigger changes that aren't ready to merge yet.

**Comment by focusaurus** [comment]
I am also seeing issues with key repeat. I'm holding arrow keys and it looks like there's significant lag between the actual registering of the key press events and updating the GUI and cursor position, so the cursor movement lags behind the back end truth and therefore when I release the key, the cursor continues to move for another 1-2 seconds.

I'm on linux under X running version wezterm 20220905-102802-7d4b8249.

I've confirmed the issue does not occur on another terminal (kitty).

Anyone know if there's still any pending key repeat work in branches? Wez's comment was like 18 months ago so that specific branch probably landed by now, but maybe I'll go reading through the git history.

**Reply by wez** [reply]
Those changes landed a long time ago, and were for Wayland. If you are using X11, that has a totally different way to receive keyboard input. Please file an issue; in addition to the questions it asks, please also include information about your key repeat rate and any other keyboard info that might be relevant. You may also wish to experiment with https://wezfurlong.org/wezterm/config/lua/config/use_ime.html

**Reply by focusaurus** [reply]
OK once I get a handle on this consistently, I'll file an issue. Right now it seems to be intermittent. Possibly load related. Unsure, but it was happening earlier and I didn't restart anything but it's not laggy anymore.
